[WIP]  Convert matrix expressions to Indexed

anpandey opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### Brief description of what is fixed or changed

Adds an `as_indexed` method that converts a matrix expression to its corresponding contraction using represented by `Indexed` objects, functioning as an inverse of `from_index_summation`.

```python
>>> n, m = symbols('n m', integer=True)
>>> A = MatrixSymbol('A', n, m)
>>> B = MatrixSymbol('B', m, n)
>>> e = A*B
>>> e.as_indexed()
A[i, j]*B[j, k]
```

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
  * add `as_indexed` to convert matrix expressions to `Indexed`
<!-- END RELEASE NOTES -->

anpandey committed:
Add as_indexed method to MatExpr

sylee957 commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -569,6 +569,84 @@ def recurse_expr(expr, index_ranges={}):
+                                                    col_idx)
+                    result.append(arg_indexed)
+
+                return Mul(*args_c, *result)
```
This may be failing in python 2. You may have to catenate the lists and pass

Upabjojr replied:
`Mul.fromiter(args_c + tuple(result))`

Upabjojr replied:
or maybe `Mul.fromiter(j for i in (args_c, result) for j in i)`

Upabjojr commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -569,6 +569,84 @@ def recurse_expr(expr, index_ranges={}):
         else:
             return remove_matelement(retexpr, first_index, last_index)
 
+    def as_indexed(self):
```
SymPy's standard is `_eval_as_Indexed` (target class named concatenated to `_eval_as_`)

anpandey replied:
Would it be the same standard if I wanted to make it externally accessible?

Upabjojr replied:
You call it with `.rewrite(Index)` ==> `rewrite` tries to look at methods like `_eval_as_Index`.

jksuom replied:
I think that should be `_eval_rewrite_as_Index` (and `_eval_rewrite_as_Indexed` for `.rewrite(Indexed)`).

Upabjojr replied:
Yes, right.

asmeurer replied:
Yes, rewrite(Indexed) is better. Not only is this a standard SymPy thing, but it will work even if the top-level expression isn't an Indexed, because every Basic object has the `rewrite` method and it automatically recurses into subexpressions.

anpandey committed:
Construct new operations using `fromiter`

anpandey committed:
Update function name to `_eval_rewrite_as_Indexed`

anpandey committed:
Fix a typo

anpandey committed:
Change helper to return row and column Idx

This is needed because the helper might not return an `Indexed`
expression, which would make the indices would not be accessible.

anpandey committed:
Add some tests

anpandey committed:
Rewrite of conversion of matrix multiplication to Indexed

czgdp1807 commented:
Looks like a good addition. @anpandey Would you still like to continue this PR?

czgdp1807 commented:
Looks like a nice addition.
